addons/delivery/models/stock_picking.py

```diff
@@ -95,6 +95,8 @@ def _add_delivery_cost_to_so(self):
         self.ensure_one()
         sale_order = self.sale_id
         if sale_order.invoice_shipping_on_delivery:
+            if self.carrier_id.delivery_type in ['fixed', 'base_on_rule']:
+                self.carrier_price = self.carrier_id.price
```

Comment:
I think the price could be changed on the picking at the moment where the carrier is modified instead of waiting for the transfer. 

Caution: if you do that with an onchange, don't change the price if the carrier is not fixed/base on rule.